pageserver/src/tenant.rs

```diff
@@ -1882,54 +1910,26 @@ impl Tenant {
             .enter()
             .map_err(|_| CreateTimelineError::ShuttingDown)?;
 
-        // Get exclusive access to the timeline ID: this ensures that it does not already exist,
-        // and that no other creation attempts will be allowed in while we are working.
-        let create_guard = match self.create_timeline_create_guard(new_timeline_id) {
```

Comment:
Is there a reason why we can't do `create_timeline_create_guard` upfront? It's currently in `Tenant::start_creating_timeline`, but calling it only requires `new_timeline_id` which we have here.

I'm in favour of grabbing such things at the entry point. Makes it easier to spot.

Reply:
Ah it's because you might not know the `ancestor_lsn` for the branching case and you want to validate it:
`we finally have determined the ancestor_start_lsn, so we can get claim exclusivity now`.

Still not sure I see it though. You could just pass the guard down and leave the idempotency checks where they are.

Reply:
> Ah it's because you might not know the ancestor_lsn for the branching case and you want to validate it:
we finally have determined the ancestor_start_lsn, so we can get claim exclusivity now.

Correct.


> Ah it's because you might not know the ancestor_lsn for the branching case and you want to validate it:
we finally have determined the ancestor_start_lsn, so we can get claim exclusivity now.

> Still not sure I see it though. You could just pass the guard down and leave the idempotency checks where they are.

The plan is to pass the creation arguments (called "CreatingTimelineState" in the version you reviewed) into the routine that acquires the creation guard, and do the idempotency check there.

=> https://github.com/neondatabase/neon/pull/9501